activestorage/CHANGELOG.md

```diff
@@ -1,3 +1,11 @@
+*   Fix `ActiveStorage::Representations::ProxyController` not returning the proper
+    preview image variant for previewable files. All `representables` ( `Variant`
+    `VariantWithRecord` and `Preview`) have now full interchangeable APIs, and
+    the use of `ActiveStorage::Blob#representation` is now recommended over
+    `ActiveStorage::Blob#variant` or `ActiveStorage::Blob#preview`.
```

Comment:
> Why would ActiveStorage::Blob#representation now be recommended over ActiveStorage::Blob#variant or ActiveStorage::Blob#preview (in the case where you know that you have one or the other)?

Because technically speaking, even if we have a `has_one_attached :pdf_file` we can **never** know whether a blob is variable or previewable or neither. At least in ActiveStorage( unless we use a third party gem like active_storage_validations)
I believe the safest way, in ActiveStorage to ask for some variant or preview is with: 
```ruby
begin 
  @blob.representation({..}) # the blob can be of any content type technically.
rescue UnrepresentableError
 # re-raise or ignore depending on the use case.
 end
 ```
 
 The main goal of this PR was to fix the wrong preview image transformation. However since it led to harmonizing with the `include Servable` (I'm summarizing), I thought I might add some info related to that in the changelog

Reply:
> I believe the safest way, in ActiveStorage to ask for some variant or preview

In terms of not raising an exception, yes.  However, if your application logic is expecting a variable (image-like) file, then you probably want to use `variant` because `representation` could mask a problem.  For example, if you have `has_one_attached :avatar`, and the user uploads a video instead of an image, then `representation` would sweep the problem under the rug.

But also, regardless of which method you prefer, I don't think any that of changed in this PR, right?  We fixed how `ProxyController` handles previews, but we didn't expose any new behavior otherwise.  `ActiveStorage::Blob::Servable` is still a private API.  Likewise, even though we made it possible to pass `Variant`, `VariantWithRecord`, or `Preview` to `send_blob_stream`, we haven't made that part of its public API.  The public API for `send_blob_stream` still requires an `ActiveStorage::Blob` instance.

Reply:
@jonathanhefner  Alright I agree.

> The public API for send_blob_stream still requires an ActiveStorage::Blob instance.

This is true. And thankfully, since our proxy tests now properly test all cases ( + not calling `#image`), if `send_blob_stream` evolves,  for example, by calling `blob.method_that_does_not_exist_on_variant`,  our tests will catch it. So this is now reassuring!

However, my concern regarding the `Preview#image` still remains. But that's for another day.

Thanks for replying!